File: 037_LongestZigZagPathInBinaryTree.cpp
// Topic: Binary Tree / DFS
// Longest ZigZag Path in a Binary Tree: longest path alternating left/right.
// Time Complexity: O(n)
// Space Complexity: O(h)

#include <bits/stdc++.h>
using namespace std;
struct TreeNode
{
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

int bestZ = 0;
pair<int, int> dfsZ(TreeNode* node)
{  // {len ending with left, len ending with right}
    if (!node)
        return {0, 0};
    auto L = dfsZ(node->left);
    auto R = dfsZ(node->right);
    int endLeft = 1 + L.second;  // previous step was right then left
    int endRight = 1 + R.first;
    bestZ = max({bestZ, endLeft, endRight});
    return {endLeft, endRight};
}

int longestZigZag(TreeNode* root)
{
    bestZ = 0;
    dfsZ(root);
    return bestZ == 0 ? 0 : bestZ - 1;
}

int main()
{
    TreeNode* r = new TreeNode(1);
    r->right = new TreeNode(1);
    r->right->left = new TreeNode(1);
    r->right->left->right = new TreeNode(1);
    r->right->left->right->left = new TreeNode(1);
    cout << longestZigZag(r) << "\n";  // example value
    TreeNode* r2 = new TreeNode(1);
    cout << longestZigZag(r2) << "\n";  // 0
    return 0;
}

File: 037_LongestZigZagPathInBinaryTree.py
"""
Topic: Tree / DFS
Longest ZigZag Path in Binary Tree: DFS tracking direction and length
Time Complexity: O(n)
Space Complexity: O(h)
"""

from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def longest_zigzag(root: Optional[TreeNode]) -> int:
    best = 0
    def dfs(node):
        nonlocal best
        if not node: return (-1, -1)  # (left_len, right_len)
        left = dfs(node.left)
        right = dfs(node.right)
        cur_left = 1 + left[1]
        cur_right = 1 + right[0]
        best = max(best, cur_left, cur_right)
        return (cur_left, cur_right)
    dfs(root)
    return best

if __name__ == '__main__':
    # example small tree
    root = TreeNode(1, TreeNode(2, None, TreeNode(3, TreeNode(4), TreeNode(5))), None)
    print(longest_zigzag(root))  # example value
    print(longest_zigzag(None))  # 0
